pallets/pooled-staking/src/traits.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright (C) Moondance Labs Ltd.
+// This file is part of Tanssi.
+
+// Tanssi is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Tanssi is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Tanssi.  If not, see <http://www.gnu.org/licenses/>
+
+use {
+    crate::{Config, Error},
+    sp_core::U256,
+    sp_runtime::traits::{CheckedAdd, CheckedMul, CheckedSub, Zero},
+    sp_std::convert::TryInto,
+};
+
+pub struct OverflowError;
+
+impl<T: Config> From<OverflowError> for Error<T> {
+    fn from(_: OverflowError) -> Self {
+        Error::MathOverflow
+    }
+}
+
+pub struct UnderflowError;
+
+impl<T: Config> From<UnderflowError> for Error<T> {
+    fn from(_: UnderflowError) -> Self {
+        Error::MathUnderflow
+    }
+}
+
+pub trait MulDiv: Sized {
+    /// Multiply self by `a` then divide the result by `b`.
+    /// Computation will be performed in a bigger type to avoid overflows.
+    /// After the division, will return `None` if the result is to big for
+    /// the real type or if `b` is zero.
+    fn mul_div(self, a: Self, b: Self) -> Result<Self, OverflowError>;
```

Comment:
There's a similar trait in substrate: [sp_arithmetic::rational::MultiplyRational](https://docs.rs/sp-arithmetic/latest/sp_arithmetic/rational/trait.MultiplyRational.html)

The implementation looks a bit different, so if you decide to use it make sure to test edge cases. For instance, `multiply_rational` returns `None` on `0 / 0` while `mul_div` returns `0`.

Reply:
Thanks will fix this edgecase. I didn't find this trait in Substrate, but we don't need the rounding complexity, only something that behaves like `a * b / c` while preventing overflow on `a * b`.